backend/src/routes/reservations.js

```diff
@@ -0,0 +1,100 @@
+const express = require('express');
+const router = express.Router();
+const auth = require('../auth');
+const db = require('../db');
+const utils = require('../utils');
+const EVENT_RESERVATION_LIMIT = process.env.EVENT_RESERVATION_LIMIT || 5;
+const getReservations = async (req, res) => {
+    try {
+        const user = req.user;
+        const dbuser = await db.users.getUser(user.phoneNumber);
+        const reservations = await db.reservations.getReservations(dbuser.id);
+        res.status(200).send(reservations);
+    } catch (err) {
+        res
+            .status(500)
+            .send(`Failed to create reservation. Error: ${err.message}`);
+    }
+};
+
+const createReservation = async (req, res) => {
+    try {
+        const reservationObj = req.body;
+        const validationErrors = utils.reservations.validateReservationRequest(
+            reservationObj
+        );
+        if (validationErrors) {
+            res.status('420').send(validationErrors);
+        } else {
+            //checking whether the user have enough saldo/balance
+            const user = await db.users.getUserBalance(reservationObj.userId);
+            const notEnoughBalance = utils.reservations.checkBalance(
+                user.balance,
+                reservationObj.ticketPrice
+            );
+
+            //checking whether already the limit for the particular event has reached.
+            const existingReservations = await db.reservations.getReservationCount(
+                reservationObj.eventId
+            );
+            const bookingLimitReached = utils.reservations.checkBookingLimit(
+                existingReservations.count,
+                EVENT_RESERVATION_LIMIT
+            );
+
+            //check whether the same event has been already reserverd by the user.
+            const event = await db.reservations.getReservationForEvent(
+                reservationObj.eventId,
+                reservationObj.userId
+            );
+
+            if (bookingLimitReached) {
+                res.status(420).send(bookingLimitReached);
+            } else if (notEnoughBalance) {
+                res.status(420).send(notEnoughBalance);
+            } else if (event) {
+                res
+                    .status(420)
+                    .send(
+                        'You have already made a reservation for the same event!. Try for another event.'
+                    );
+            } else {
+                await db.reservations.createReservation(reservationObj);
```

Comment:
We should ensure the updates for reservations and balance are ran in a single transaction.